### PR TITLE
[AMDGPU] Accept arbitrary sized sources in CalculateByteProvider

### PR DESCRIPTION
**Pull request by jrbyrnes**
Reland the original patch with additional commit containing fix for two issues:

1. Attempting to bitcast using MVTs with no corresponding LLVM type. getDWordFromOffset now works directly with the original vector to get the corresponding elements given the DWordOffset.
2. Improper bit tracking in CalculateByteProvider for vector types using certain ops. Previously, bit tracking for certain ops (e.g. ISD::TRUNCATE) assumed operands were scalar types, which is not correct since these ops have different semantics depending on vector / scalar. CalculateByteProvider / CalculateSrcByte now exit on vector types, handling which is a TODO.

The patch containing the fixes has not been reviewed yet -- so I've separated out the commits to make reviewing easier. Will land them atomically.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -10989,6 +10996,7 @@ calculateByteProvider(const SDValue &Op, unsigned Index, unsigned Depth,
 
   case ISD::SRA:
   case ISD::SRL: {
+    if (IsVec) return std::nullopt;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:10999]: new line

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -11014,6 +11022,7 @@ calculateByteProvider(const SDValue &Op, unsigned Index, unsigned Depth,
   }
 
   case ISD::SHL: {
+    if (IsVec) return std::nullopt;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:11025]: new line

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 3]
```diff
@@ -11038,6 +11047,7 @@ calculateByteProvider(const SDValue &Op, unsigned Index, unsigned Depth,
   case ISD::SIGN_EXTEND_INREG:
   case ISD::AssertZext:
   case ISD::AssertSext: {
+    if (IsVec) return std::nullopt;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:11050]: new line

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 4]
```diff
@@ -11068,6 +11078,7 @@ calculateByteProvider(const SDValue &Op, unsigned Index, unsigned Depth,
   }
 
   case ISD::TRUNCATE: {
+    if (IsVec) return std::nullopt;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:11081]: new line

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 5]
```diff
@@ -11114,9 +11125,11 @@ calculateByteProvider(const SDValue &Op, unsigned Index, unsigned Depth,
     return std::nullopt;
   }
 
-  case ISD::BSWAP:
+  case ISD::BSWAP: {
+    if (IsVec) return std::nullopt;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:11129]: new line

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 6]
```diff
@@ -11133,6 +11146,7 @@ calculateByteProvider(const SDValue &Op, unsigned Index, unsigned Depth,
   }
 
   case AMDGPUISD::PERM: {
+    if (IsVec) return std::nullopt;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:11149]: new line

**Comment by jrbyrnes** [comment]
Latest iteration passes PSDB

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 7]
```diff
@@ -11233,15 +11259,68 @@ static bool hasNon16BitAccesses(uint64_t PermMask, SDValue &Op,
   return !addresses16Bits(Low16) || !addresses16Bits(Hi16);
 }
 
+static SDValue getDWordFromOffset(SelectionDAG &DAG, SDLoc SL, SDValue Src,
+                                  unsigned DWordOffset) {
+  SDValue Ret;
+
+  auto ValueSize = Src.getValueSizeInBits().getFixedValue();
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:11266]: TypeSize

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 8]
```diff
@@ -11233,15 +11259,68 @@ static bool hasNon16BitAccesses(uint64_t PermMask, SDValue &Op,
...
+  // ByteProvider must be at least 8 bits
+  assert(!(ValueSize % 8));
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:11268]: .isKnownMultipleOf(8)

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 9]
```diff
@@ -11302,8 +11379,16 @@ static SDValue matchPERM(SDNode *N, TargetLowering::DAGCombinerInfo &DCI) {
       return DAG.getBitcast(MVT::getIntegerVT(32), Op);
   }
 
+  SDValue OtherOp =
+      SecondSrc.has_value() ? *PermNodes[SecondSrc->first].Src : Op;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:11383]: Don't need has_value

**Comment by jrbyrnes** [comment]
Rebase + Ping

**Review by arsenm — APPROVED** [review]

**Pull request merged**